src/aeso_ast_infer_types.erl

```diff
@@ -845,10 +861,11 @@ infer1(Env, [{Contract, Ann, ConName, Code} | Rest], Acc, Options)
         contract -> ets_insert(defined_contracts, {qname(ConName)});
         contract_interface -> ok
     end,
-    {Env1, Code1} = infer_contract_top(push_scope(contract, ConName, Env), What, Code, Options),
+    Env0 = bind_contract(untyped, Contract0, Env),
+    {Env1, Code1} = infer_contract_top(push_scope(contract, ConName, Env0), What, Code, Options),
     Contract1 = {Contract, Ann, ConName, Code1},
     Env2 = pop_scope(Env1),
-    Env3 = bind_contract(Contract1, Env2),
+    Env3 = bind_contract(typed, Contract1, Env2),
```

Comment:
Do you really need to rebind it? I would expect the "untyped" version to unify to the final "typed" representation anyway. Do you check if they match?

Reply:
If I don't rebind, the cases with type aliases will fail, since they are only replaced during the type inference.

Reply:
Could you give me some example? I also wonder if this could be hacked around somehow (that is, so both versions don't really match)

Reply:
This is one example:
```
contract interface Remote =
  record rstate = { i : int, s : string, m : map(int, int) }

  entrypoint fun_update_i  : (rstate, int) => rstate

contract StateHandling =

  type state = Remote.rstate

  entrypoint init(r : Remote, i : int) =
    let state0 = { i = 0, s = "undefined", m = {} }
    r.fun_update_i(state0, i)
```

The problem here is that the type of the arguments in the `fun_update_i` function declaration is set to `rstate` before the type inference and to `Remote.rstate` after the type inference (because the qualifications are added during type inference).

Binding the fields only before the type inference, would cause an error that the types `rstate` and `Remote.rstate` cannot be unified.

I can't currently think of a solution for this, so I may add a comment about it and maybe it can be handled later, unless you can think of a solution for this?

Reply:
Maybe forcefully qualify all the types at every opportunity, so there is always `Remote.rstate`?